### PR TITLE
Fix hit error meters not updating visual state when hidden

### LINKED ISSUE #18624: Hit Error Bar does not update when hidden
### Type

Cosmetic

### Bug description

The hit error bar does not update while it is hidden – meaning it does not add any lines and does not move its ‘caret’ (which shows the average).
I would expect all of this to happen in the background, even while the bar is not shown to the player.

When the HUD is then shown (i.e. during a map's break or when pausing), _sometimes_ all the past lines get added all at once, making the bar look crowded and full, and _sometimes_ they just never show up, meaning the bar is completely empty even after having played a part of the map.
I feel like this is based on the time after which the bar is shown, but I could be wrong on that. From my testing, if the bar was shown relatively early (say, at most 1 minute after map start?), the lines don't show up at all, if it takes longer for the bar to show up they get shown all at once.

I suspect that this is related to #15568, which talks about the UR counter, not the Hit Error Bar. The behavior seems to be pretty much the same, but it could still be totally unrelated, I haven't looked into it.

## Example of the hit error bar not updating
osu_2022-06-08_17-39-27
## What it should look like at that point
Simulated by watching the replay with the HUD set to always-on.
Note that there are lines and the caret is slightly off to the left, as I've been hitting a bit early.
osu_2022-06-08_17-53-10
## Example of many line appearing all at once, later on in the map
As shown by the difficulty indicator at the bottom, there weren't a lot of circles right before the break or anything.
osu_2022-06-08_17-34-41

### Version

2022.605.1

### Logs

[database.log](https://github.com/ppy/osu/files/8863239/database.log)
[network.log](https://github.com/ppy/osu/files/8863240/network.log)
[performance-audio.log](https://github.com/ppy/osu/files/8863241/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/8863242/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/8863243/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/8863244/performance-update.log)
[performance.log](https://github.com/ppy/osu/files/8863245/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8863246/runtime.log)

### PR DESCRIPTION
**Pull request by peppy**
It is an expectation of users that when the HUD is shown after a period of being hidden, it will visually reflect the state based on recent judgements.

To achieve this, I've added `AlwaysPresent` and moved the transform application to the meter level, rather than at a child level. If this is seen as a bad direction, `AlwaysPresent` can be applied to the drawable children and the transforms can be moved back.

Also of note, `ColourHitErrorMeter` is pretty weird. The flow class could potentially be removed and reduce `AlwaysPresent` usage by one. Can do that refactor as part of this PR if preferred.

Closes #18624.

### CODE REVIEW

File: osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs [hunk 1]
```diff
@@ -420,12 +423,26 @@ protected override void OnNewJudgement(JudgementResult judgement)
...
                 Colour = GetColourForHitResult(judgement.Type),
-            });
+                Alpha = 0,
+                Width = 0,
+            };
+
+            judgementsContainer.Add(judgementLine);
+
+            // Importantly, transforms should be applied in this method rather than constructed drawables
+            // to ensure that they are applied even when the `HitErrorMeter` is hidden (see `AlwaysPresent` usage).
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs:438]: I may have been doing something wrong in the earlier tests, I'll revisit this. It looked initially like children weren't being `LoadComplete`d even though the `HitErrorMeter` was set to `AlwaysPresent`.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs]: Reverted all unnecessary changes. Some of them felt like nice clean-up but probably best to keep this focused and simple.

**Comment by frenzibyte** [comment]
While this fixes `HitErrorMeter` not updating when hidden by itself, it still doesn't update when combined with the `HUDOverlay`, and which I confirmed via logging on `HitErrorMeter.Update()` and going to a player session:

https://user-images.githubusercontent.com/22781491/173461407-0e839b01-73fe-479c-bddf-a64aa3be05be.mp4

Reason would be that the layer which hides in HUD overlay is the `MainHUDComponents` container, and once it's hidden and `IsPresent = false`, its children (HUD components, including `HitErrorMeter`) [cannot possibly update](https://github.com/ppy/osu-framework/blob/7610d868ead7f86e2606ba6b3f23b3c5b2605648/osu.Framework/Graphics/Containers/CompositeDrawable.cs#L906-L908).

Easiest fix I can think of would be adding `AlwaysPresent` to the HUD components container, but would imagine the components hiding individually in the future to avoid this, maybe.

**Review by frenzibyte — CHANGES_REQUESTED** [review]
As noted.

**Comment by peppy** [comment]
I tested in gameplay and it worked for me correctly. Maybe a regression with the last commit, try without that?

**Comment by frenzibyte** [comment]
Can still see the issue after reverting last commit:

https://user-images.githubusercontent.com/22781491/173472699-8d5f6baa-0324-45b7-a7c9-8db82f8597cf.mp4

I believe this sort of issue may fix itself probably by having `HitErrorMeter` [not rely on scheduling](https://github.com/ppy/osu/blob/0771265cafbd0db82403f7e95cff892906a1bbfc/osu.Game/Screens/Play/HUD/HitErrorMeters/HitErrorMeter.cs#L46-L47) (which should be the reason why all judgements queue up to appear at the point of displaying meter again). Haven't tested that out myself, but just putting it here for visibility.

**Comment by peppy** [comment]
Are you actually *looking at the display*? It doesn't matter if update is running or not, the important thing is whether it behaves correctly.

**Comment by peppy** [comment]
I've gone with the `AlwaysPresent` solution for now as it's the simplest, and will potentially cover other HUD elements that require similar updates.

Ensure containers which are being used to hide HUD elements still update for now

I don't think this is necessarily a final solution (as this means all HUD elements are adding overhead even when not visible), but this will make the implementations much easier for the time being.

I've checked and can't notice any perceivable overhead in profiling so we should be fine for now.

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 2]
```diff
@@ -108,6 +111,7 @@ public HUDOverlay(DrawableRuleset drawableRuleset, IReadOnlyList<Mod> mods)
                     Margin = new MarginPadding(10),
                     Spacing = new Vector2(10),
                     AutoSizeAxes = Axes.Both,
+                    AlwaysPresent = true,
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUDOverlay.cs:114]: Yes, you are correct in thinking that.

**Review by bdach — APPROVED** [review]

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**